### DISCUSSION TITLE
MoltenVK/Metal timeout issue

### DISCUSSION THREAD

**Discussion by RevoluPowered** [KhronosGroup/MoltenVK#1748] [Q&A]
Basically I'm having a timeout issue with a project:
- https://github.com/godotengine/godot/issues/67373

```css
2022-10-14 00:20:24.210364+0100 godot.macos.opt.tools.x86_64[4888:74403] Execution of the command buffer was aborted due to an error during execution. Caused GPU Timeout Error (00000002:kIOAccelCommandBufferCallbackErrorTimeout)
[mvk-error] VK_ERROR_DEVICE_LOST: Command buffer 0x7fe954f59a00 "vkQueueSubmit CommandBuffer on Queue 0-0" execution failed (code 2): Caused GPU Timeout Error (00000002:kIOAccelCommandBufferCallbackErrorTimeout)
2022-10-14 00:20:24.210533+0100 godot.macos.opt.tools.x86_64[4888:74403] ERROR:  - Message Id Number: 0 | Message Id Name: 
	VK_ERROR_DEVICE_LOST: Command buffer 0x7fe954f59a00 "vkQueueSubmit CommandBuffer on Queue 0-0" execution failed (code 2): Caused GPU Timeout Error (00000002:kIOAccelCommandBufferCallbackErrorTimeout)
	Objects - 1
		Object[0] - VK_OBJECT_TYPE_QUEUE, Handle 105553183017928
at: _debug_messenger_callback (drivers/vulkan/vulkan_context.cpp:171)
ERROR:  - Message Id Number: 0 | Message Id Name: 
	VK_ERROR_DEVICE_LOST: Command buffer 0x7fe954f59a00 "vkQueueSubmit CommandBuffer on Queue 0-0" execution failed (code 2): Caused GPU Timeout Error (00000002:kIOAccelCommandBufferCallbackErrorTimeout)
	Objects - 1
		Object[0] - VK_OBJECT_TYPE_QUEUE, Handle 105553183017928
```

What I've tried:
- running ubsan/tsan/lsan
- xcode frame debugger
- setting various flags in MVK 

```css
MKV_DEBUG=1;
MTL_SHADER_VALIDATION=1;
MVK_ALLOW_METAL_EVENTS=1;
MVK_CONFIG_LOG_LEVEL=3
```

Enabling shader validation in xcode prevents the application killing my operating system, so I am curious, is there additional options I can enable to get the shader validation information out of moltenvk? is there some flags or something I can enable to print metal errors etc from moltenvk? 

I believe the issue is we are passing some bad data to a compute shader or vice versa, but I can't diagnose because nothing obvious jumps out even after running the xcode frame debugger.

What I want is the errors from metal, if they are hidden behind some configuration option, etc, or a way of listing all the bad memory accesses to the log.

**Comment by billhollings** [comment]
Hmmm...we already log all the info that comes back from Metal, so we may not be able to get more specific than that.

With your settings including `MKV_DEBUG=1` and `MVK_CONFIG_LOG_LEVEL=3`, you should be seeing the SPIR-V->MSL shader conversions being logged. Assuming the shaders are being converted just after they are first being executed in a pipeline, you might be able to isolate the shader that is stalling and causing the GPU timeout (assuming it's a shader that is stalling).

The fact that `MTL_SHADER_VALIDATION` is set, and you're not seeing a shader validation error, might indicate that the shader is valid, but still hits a stalled loop, possibly? 

If you can isolate the shader, pull the SPIR-V, MSL, and reconstructed GLSL from the MoltenVK log, and post it here, we might be able to determine if we have an error during shader conversion that generates a stalled loop, or possibly isolate the loop that might be stalled by bad incoming data, as you indicate.
